modules/xfd.ts

```diff
@@ -0,0 +1,2218 @@
+		Window.addFooterLink('Twinkle help', 'WP:TW/DOC#xfd');
+		this.makeForm(Window);
+	}
+
+	// invoked only once
+	makeForm(Window) {
+		this.Window = Window;
+		let form = new Morebits.quickForm(() => { this.mode.evaluate(); });
+
+		form.append({
+			type: 'select',
+			name: 'venue',
+			label: 'Deletion discussion venue:',
+			tooltip: 'When activated, a default choice is made, based on what namespace you are in. This default should be the most appropriate.',
+			event: this.onCategoryChange.bind(this),
+			list: Xfd.modeList.map((mode) => ({
+				type: 'option',
+				label: mode.venueLabel,
+				selected: this.mode instanceof mode,
+				value: mode.venueCode
+			}))
+		});
+
+		form.append({
+			type: 'div',
+			id: 'wrong-venue-warn',
+			style: 'color: red; font-style: italic'
+		});
+
+		form.append({
+			type: 'checkbox',
+			list: [
+				{
+					label: 'Notify page creator if possible',
+					value: 'notify',
+					name: 'notifycreator',
+					tooltip: "A notification template will be placed on the creator's talk page if this is true.",
+					checked: true
+				}
+			]
+		});
+
+		this.fieldset = form.append({
+			type: 'field',
+			label: 'Work area',
+			name: 'work_area'
+		});
+
+		var previewlink = document.createElement('a');
+		$(previewlink).click(() => {
+			this.mode.preview(this.result);  // |result| is defined below
+		});
+		previewlink.style.cursor = 'pointer';
+		previewlink.textContent = 'Preview';
+		form.append({ type: 'div', id: 'xfdpreview', label: [ previewlink ] });
+		form.append({ type: 'div', id: 'twinklexfd-previewbox', style: 'display: none' });
+
+		form.append({ type: 'submit' });
+
+		this.result = form.render();
+		Window.setContent(this.result);
+		Window.display();
+		this.result.previewer = new Morebits.wiki.preview(document.getElementById('twinklexfd-previewbox'));
+
+		// We must init the controls
+		var evt = document.createEvent('Event');
+		evt.initEvent('change', true, true);
+		this.result.venue.dispatchEvent(evt);
+
+		return form;
+	}
+
+	// invoked on every mode change
+	onCategoryChange(evt) {
+		var venueCode = evt.target.value;
+		var form = evt.target.form;
+
+		let mode = Xfd.modeList.filter((mode) => {
+			return mode.venueCode === venueCode;
+		})[0];
+		if (!mode) {
+			throw new Error('Unrecognized venue: ' +  venueCode); // should never happen
+		}
+		// @ts-ignore
+		this.mode = new mode();
+		this.mode.result = this.result;
+		this.mode.Window = this.Window;
+
+		$('#wrong-venue-warn').text(this.mode.getVenueWarning() || '');
+		form.previewer.closePreview();
+
+		let fieldset = this.mode.generateFieldset();
+		let renderedFieldset = fieldset.render();
+		$(this.result).find('fieldset[name=work_area]')
+			.replaceWith(renderedFieldset);
+		this.mode.postRender(renderedFieldset as HTMLFieldSetElement);
+	}
+}
+
+
+abstract class XfdMode {
+	static venueCode: string
+	static venueLabel: string
+
+	// must be overridden, unless the venue is never the default choice
```

Comment:
"unless..." means that it isn't that this method "**must** be overridden", only that it generally should be